### PR TITLE
Fix sparse y

### PR DESCRIPTION
**Pull request by eddiebergman**
As noted in openml/automlbenchmark#370 and #1210, autosklearn failed to process sparse labels. This PR addresses that by manually coverting sparse `y` into a dense `y` before using sklearn's `type_of_target` to get the type of target for the task.

# Changes
* Introduces a check before any `type_of_target` is called to ensure the `y` data is converted to dense if it is not
* Moves the `type_of_target` checking in `AutoMLClassifier` and `AutoMLRegrssor` to `AutoML` to reduce duplicaiton
* Default metric for a given task is now located in `metrics/__init__.py` and looked up in `AutoML.fit` instead of passed down from the subclasses `AutoMLClassifer` and `AutoMLRegressor`
* Introduces an overly comprehensive test for all combinations of input types

### CODE REVIEW

File: autosklearn/data/validation.py [hunk 1]
```diff
@@ -12,6 +14,21 @@
 from autosklearn.util.logging_ import get_named_client_logger
 
 
+def convert_if_sparse(y: SUPPORTED_TARGET_TYPES) -> SUPPORTED_TARGET_TYPES:
+    # Densify sparse y
```
Comment (mfeurer) [inline autosklearn/data/validation.py:18]: Could you please write this as a proper docstring?

File: autosklearn/data/validation.py [hunk 2]
```diff
@@ -12,6 +14,21 @@
 from autosklearn.util.logging_ import get_named_client_logger
 
 
+def convert_if_sparse(y: SUPPORTED_TARGET_TYPES) -> SUPPORTED_TARGET_TYPES:
```
Comment (mfeurer) [inline autosklearn/data/validation.py:17]: Don't we need a new data type annotation for the target which isn't sparse any more?

File: autosklearn/automl.py [hunk 3]
```diff
@@ -429,11 +429,19 @@ def _do_dummy_prediction(self, datamanager: XYDataManager, num_run: int) -> int:
...
+    @classmethod
+    def _supports_task_type(cls, task_type: str) -> bool:
+        raise NotImplementedError
+
     def fit(
         self,
         X: SUPPORTED_FEAT_TYPES,
         y: SUPPORTED_TARGET_TYPES,
-        task: int,
+        task: Optional[int] = None,
```
Comment (mfeurer) [inline autosklearn/automl.py:444]: Is passing the task type still necessary? It appears that in the calls of the subclass you stopped passing the task type to the super class.

File: autosklearn/data/validation.py (hunk 2 above)
Comment (eddiebergman) [reply autosklearn/data/validation.py:17]: Yes we do, i suppose using 
* For types passed around internally `SUPPORTED_TARGET_TYPES = Union[List, Series, DataFrame, np.ndarray]`
* For external api points, the type is `Union[spmatrix, SUPPORTED_TARGET_TYPES]`

File: autosklearn/automl.py (hunk 3 above)
Comment (eddiebergman) [reply autosklearn/automl.py:444]: The tests in`test/test_automl.py` rely on creating instances of `AutoML`. In this case a `task` must be passed in.

Long story short, I tried to remove it with a workaround for the problem described below but tests started failing and the solution to remove it is a bigger PR, the goal was jsut to fix sparse y.

#### Problem
`AutoML` can't determine the task type for itself. The problem is that `type_of_target(y)` can return `'multiclass'` which we map to either `REGRESSION` or `MULTICLASS_CLASSIFICATION`, depending on the subclass. Hence, when the object is not a subclass, `AutoML` needs this information passed in.

You're right though, it's not really needed, especially not `task` and `is_classification` parameters. The minimal amount of information needed is simply `is_classification: bool`. This also removes the need for subclasses at all.
* `AutoML` can uniquely determine the task using `type_of_target` and `is_classification`
* This would mean the subclasses do nothing except `AutoMLClassifier` which does extra work in `predict` to transform predictions to labels
* This transformation could happen in `AutoSklearnClassifier`, leaving `AutoML` to simply generate outputs, removing the need for having subclasses of `AutoML`

Comment (mfeurer) [reply autosklearn/automl.py:444]: Got it, so for now we can leave everything as it is? But could you please add a docstring on why this is necessary?

Comment (eddiebergman) [reply autosklearn/automl.py:444]: Yes, I say we leave it as is and we can talk about this next week. I've added a docstring to the function with the explanation included.

**Review by mfeurer — APPROVED** [review]
This looks good now, thanks a lot. However, due to merging #1217 there are now conflicts. Could you please have a look into them?

**Comment by eddiebergman** [comment]
> This looks good now, thanks a lot. However, due to merging #1217 there are now conflicts. Could you please have a look into them?

Fixed the conflicts and running the tests, mainly just imports and a fight over which test wants to occupy the bottom of the test_automl.py -_-

**Review by mfeurer — APPROVED** [review]
